src/assets/scripts/client/aircraft/FlightManagementSystem/Fms.js

```diff
@@ -304,1237 +234,687 @@ export default class Fms {
         return waypoints[0];
     }
 
-    // TODO: this should move to a class method
     /**
-     * Returns a flattened array of each `WaypointModel` in the flightPlan
-     *
-     * This is used only in the `CanvasController` when drawing the projected
-     * aircraft path.
+     * Get the next waypoint in the flight plan, if it exists
      *
-     * Using a getter here to stay in line with the previous api.
+     * @for Fms
+     * @property nextWaypoint
+     * @type {WaypointModel}
+     */
+    get nextWaypoint() {
+        return this._routeModel.nextWaypoint;
+    }
+
+    /**
+     * Return an array of all waypoints in all legs of the route
      *
+     * @for Fms
      * @property waypoints
      * @type {array<WaypointModel>}
      */
     get waypoints() {
-        const waypointList = _map(this.legCollection, (legModel) => {
-            return [
-                ...legModel.waypointCollection
-            ];
-        });
-
-        return _flatten(waypointList);
+        return this._routeModel.waypoints;
     }
 
+    // ------------------------------ LIFECYCLE ------------------------------
+
     /**
-     * Initialize the instance and setup initial properties
+     * Initialize class properties
      *
      * @for Fms
      * @method init
      * @param aircraftInitProps {object}
+     * @param navigationLibrary {NavigationLibrary}
+     * @chainable
      */
-    init({ altitude, category, model, route }, initialRunwayAssignment) {
-        this._setCurrentPhaseFromCategory(category);
-        this._setInitialRunwayAssignmentFromCategory(category, initialRunwayAssignment);
-
-        this.flightPlanAltitude = altitude;
+    init({ altitude, category, destination, model, nextFix, origin, routeString }, navigationLibrary) {
+        this._navigationLibrary = navigationLibrary;
+        this._routeModel = new RouteModel(this._navigationLibrary, routeString);
 
-        if (category === FLIGHT_CATEGORY.DEPARTURE) {
-            this.flightPlanAltitude = model.ceiling;
-        }
+        this._verifyRouteContainsMultipleWaypoints();
+        this._initializeFlightPhaseForCategory(category);
+        this._initializeDepartureAirport(origin);
+        this._initializeDepartureRunway();
+        this._initializeArrivalAirport(destination);
+        this._initializeArrivalRunway();
+        this._initializeFlightPlanAltitude(altitude, category, model);
+        this._initializePositionInRouteToBeginAtFixName(nextFix, category);
 
-        this.legCollection = this._buildLegCollection(route);
+        return this;
     }
 
     /**
-     * Destroy the instance and reset properties
+     * Reset class properties
      *
      * @for LegModel
-     * @method destroy
+     * @method reset
+     * @chainable
      */
-    destroy() {
-        this._navigationLibrary = null;
-        this._previousRouteSegments = [];
+    reset() {
```

Comment:
Why the name change here?  We’ve reserved `.reset()` for now pooled classes. Is the Fms now pooped too?

Reply:
Ideally I'd think just about everything should be pooled, especially the classes that are reused in volume (WaypointModel, LegModel, RouteModel, Fms, AircraftModel). And I did not make the connection of why some were destroy and some were reset but since they reset the class (and I would assume destroy to also delete the instance), I just preferred the `reset` name. Also it didn't hurt that `reset` comes after `init` when `destroy` comes before `init` when you sort them. :smirk: